utilities/scripts/python/diff.py

```diff
@@ -0,0 +1,162 @@
+                color = "green" if diff.add_or_remove_flag else "red"
+                minus_or_plus = "+" if diff.add_or_remove_flag else "-"
+                obj_name = (
+                    obj_name_to_add if diff.add_or_remove_flag else obj_name_to_remove
+                )
+
+                result += "\n------\n"
+                result += add_color(obj_name, "yellow")
+                result += f"{path}\n"
+                result += add_color(f"{minus_or_plus}{ident} {diff.val}", color)
+
+            if isinstance(diff, dict):
+                for key in diff:
+                    result += impl(
+                        diff[key],
+                        obj_name_to_add,
+                        obj_name_to_remove,
+                        ident + " ",
+                        path + "\n" + f"{ident}  {key}",
+                    )
+
+            if isinstance(diff, tuple):
+                for val in diff:
+                    result += impl(
+                        val, obj_name_to_add, obj_name_to_remove, ident, path
+                    )
+
+            return result
+
+        return impl(self.diff, obj_name_to_add, obj_name_to_remove)
+
+
+@dataclass
+class Diff:
+    """
+    Calculates the difference between two dictionaries.
+    """
+
+    a: dict
+    b: dict
+    strict: bool
+
+    def get_diff(self) -> DiffResult:
+        if self.strict:
+            return self._strict_diff_()
+        else:
+            return self._inclusion_diff_()
+
+    def _inclusion_diff_(self, reverse: bool = False) -> DiffResult:
+        """
+        Calculates an inclusion diff between the 'a' and 'b' dicts in a recursive manner.
+        Checks that 'a' has an inclusion of 'b'.
+        Args:
+            reverse (bool): An inclusion diff calculation flag.
+                If 'True' evaluates that 'a' as an inclusion of 'b'.
+                If 'False' evaluates that 'b' as an inclusion of 'a'.
+        """
+
+        def impl(expected: dict, provided: dict) -> dict:
+            diff = {}
+            for key in expected:
+                if key not in provided:
+                    diff[key] = DiffEntry(expected[key], True)
+                elif not isinstance(expected[key], dict):
+                    if expected[key] != provided[key]:
+                        diff[key] = (
+                            DiffEntry(expected[key], True),
+                            DiffEntry(provided[key], False),
+                        )
+                else:
+                    res = impl(expected[key], provided[key])
+                    if res != {}:
+                        diff[key] = res
+            return diff
+
+        diff = impl(self.a, self.b) if not reverse else impl(self.a, self.b)
+        return DiffResult(diff)
+
+    def _strict_diff_(self) -> DiffResult:
+        """
+        Calculate the strict diff between  the expected and provided inputs.
+        """
+
+        def change_flags(diff):
```

Comment:
Who is calling this?  I don't see any reference to it?

Reply:
Yea, good catch, I have missed to call this function again during the refactoring